File: LeetCode/015. 3Sum/solution.py
# class Solution:
#     def threeSum(self, nums: List[int]) -> List[List[int]]:

        
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        results = []
        def dfs(idx, N, t, result):
            if N == 2:
                left, right = idx, len(nums)-1
                while left < right:
                    if nums[left]+nums[right] == t:
                        results.append(result+[nums[left],nums[right]])
                        left += 1
                        right -= 1
                        while left < right and nums[left] == nums[left-1]: left += 1
                        while left < right and nums[right] == nums[right+1]: right -= 1
                        
                    elif nums[left]+nums[right] > t:
                        right -= 1
                    else:
                        left += 1
                
                return
               
            for i in range(idx,len(nums)-N+1):
                if t < nums[i]*N or t > nums[-1]*N:
                    break
                if i == idx or nums[i] != nums[i-1]:
                    dfs(i+1, N-1, t-nums[i], result+[nums[i]])     
            return
        dfs(0,3,0,[])
        return results
                    
            

class Solution:
    def threeSum(self, nums):
        ht={}
        nums.sort()
        ans=set()
        for i,n in enumerate(nums):
            ht[n]=i
        for i in range(len(nums)):
            for j in range(i+1,len(nums)):
                target=-(nums[i]+nums[j])
                if target in ht and ht[target]>j:
                    ans.add((nums[i],nums[j],-(nums[i]+nums[j])))
                    
        return list(ans)        

class Solution:
    def threeSum(self, nums):
        nums.sort()
        res = []
        mp = {v: i for i, v in enumerate(nums)}
        for i in range(len(nums)):
            if i and nums[i] == nums[i-1]:
                continue
            for j in range(i+1,len(nums)):
                if j > i+1 and nums[j] == nums[j-1]:
                    continue
                target=-(nums[i]+nums[j])
                if target in mp and mp[target]>j:
                    res.append([nums[i],nums[j],target])
        return res

File: LeetCode/015. 3Sum/solution.cpp
// class Solution {
// public:
//     vector<vector<int>> threeSum(vector<int>& nums) {
//         vector<vector<int>> res;
//         unordered_set<int> mp;
//         int n = nums.size();
//         sort(nums.begin(), nums.end());
//         for (int i = 0; i < n; i++){
//             for (int j = i+1; j < n; j++){
//                 if (mp.count(-nums[i]-nums[j])){
//                     res.push_back({-nums[i]-nums[j],nums[i],nums[j]});
//                 }
//             }
//             mp.insert(nums[i]);
//         }
//         if (res.empty()) return {};
//         sort(res.begin(), res.end());
//         int j = 1;
//         for (int i = 1; i < res.size(); i++){
//             if (res[i] != res[i-1]) res[j++] = res[i];
//         }
//         res.resize(j);
//         return res;
//     }
// };


class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        unordered_map<int,int> mp;
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for (int i = 0; i < n; i++) mp[nums[i]] = i;
        for (int i = 0; i < n; i++){
            if (i && nums[i] == nums[i-1]) continue;
            for (int j = i+1; j < n; j++){
                if (j > i+1 && nums[j] == nums[j-1]) continue;
                int target = -nums[i]-nums[j];
                if (mp.count(target) && mp[target] > j){
                    res.push_back({nums[i],nums[j],target});
                }
            }
        }
        return res;
    }
};



class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int n = nums.size();
        for (int i = 0; i < n && nums[i] <= 0; i++){

            if (i && nums[i] == nums[i-1])
                continue;
            for (int left = i+1, right = n-1; left < right;){
                if (nums[left] + nums[right] > -nums[i])
                    right--;
                else if (nums[left] + nums[right] < -nums[i])
                    left++;
                else{
                    res.push_back({nums[i],nums[left],nums[right]});
                    while (left+1 < right && nums[left+1] == nums[left])
                        left++;
                    left++;
                    while (left < right-1 && nums[right-1] == nums[right])
                        right--;
                    right--;
                }
            }
        }
        return res;
    }
};

